### PR TITLE
GFS components update for hpc-stack support

### LINKED ISSUE #164: Updating global workflow to use hpc-stack
The develop branch of ufs-weather-model now uses hpc-stack for all the libraries. 

The following changes need to be made to the global workflow
  1. The codes that are in the global workflow repository should now use hpc-stack
  2. The repositories that global workflow connects to also need to make updates to hpc-stack

We should track the PRs for all the component repositories through here. (This is for the top of develop)

Checklist for global-workflow external components:

- [x] Component uses stack: FV3/ufs-weather-model
- [x] Component uses stack: GSI (Hera only, waiting for other platform updates)
- [x] Component uses stack: GLDAS
- [x] Component uses stack: UPP
- [x] Component uses stack: UFS_UTILS
- [x] Component uses stack: EMC_gfs_wafs

Checklist for non-component/machine-specific stack updates (e.g. module_base, machine-setup.sh):
- [x] Hera-specific changes
- [x] Orion-specific changes
- [x] WCOSS-Dell-specific changes

### PR DESCRIPTION
**Pull request by KateFriedman-NOAA**
Update system components to versions that now support hpc-stack:

- ufs-weather-model @ [9350745](https://github.com/ufs-community/ufs-weather-model/tree/9350745855aebe0790813e0ed2ba5ad680e3f75c)
- GSI @ [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6)
- GLDAS @ [gldas_gfsv16_release.v1.15.0](https://github.com/NOAA-EMC/GLDAS/releases/tag/gldas_gfsv16_release.v1.15.0)
- UFS_UTILS @ ufs_utils_1_4_0 (already in develop)
- EMC_post @ [ff42e02](https://github.com/NOAA-EMC/EMC_post/tree/ff42e0227d6100285d4179a2572b700fd5a959cb)
- EMC_gfs_wafs @ [c2a29a6](https://github.com/NOAA-EMC/EMC_gfs_wafs/tree/c2a29a67d9432b4d6fba99eac7797b81d05202b6)
- TC_tracker @ [TC_tracker.v1.1.15.4](https://github.com/NOAA-EMC/TC_tracker/releases/tag/TC_tracker.v1.1.15.4)

EMC_verif-global version not updating in this PR.

Workflow updates:

- update FIX_DIR paths in link_fv3gfs.sh to point to developmental set instead of frozen GFSv16 set (fix_nco_gfsv16 -> fix)
- update module loads to use hpc-stack (hpc/1.1.0) on supported platforms
- update obsproc pointers to versions on disk updated for hpc-stack
- set gfs_util build to run on all machines (stack build update)
- update build and link scripts to ingest component build/link step updates
- update build_fv3.sh to build CCPP version without waves (WW3 build updates not yet ready, disabling waves for now)
- ~~set DO_WAVE=NO by default for now~~ already committed with issue #178 PR
- ~~set RUN_CCPP=YES by default~~ already committed with issue #178 PR

Tested on WCOSS-Dell, Hera, and Orion. The feature/hpc-stack branch is also being used in port to Comet.

Close #164

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
@KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?

**Comment by RussTreadon-NOAA** [comment]
hpc-stack entered the [NOAA-EMC/GSI](https://github.com/NOAA-EMC/GSI) [master](https://github.com/NOAA-EMC/GSI/tree/master) at [9c1fc15](https://github.com/NOAA-EMC/GSI/commit/9c1fc15d42573b398037319bbf8d5143ad126fb6).   The master has been updated several times since 9c1fc15.  Any NOAA-EMC/GSI master revision after the 9c1fc15 commit, completed  3/9/202, should work with hpc-stack.   @MichaelLueken-NOAA is the NOAA-EMC/GSI code manager.

**Comment by WenMeng-NOAA** [comment]
@KateFriedman-NOAA That's EMC_post version should be working for FV3GFS offline post except wafs dataset generation. @YaliMao-NOAA  has been working on that issue.  I saw the version of  ufs-weather-model still uses old upp library 10.0.0. If @junwang-noaa provides you an up-to-date version, I might give you a new version of EMC_post for matching.

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?

Thank you for catching my mistake. That hash should have been 9d9dd23e37a51fcc5ff6b7499b834c85ab32e5f3 (develop branch with beta stack version). I also tested with your feature/stack branch. I'd like to submit this PR in the next week. Thanks!

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA That's EMC_post version should be working for FV3GFS offline post except wafs dataset generation. @YaliMao-NOAA has been working on that issue. 

@WenMeng-NOAA Gotcha...I'm also waiting for a version of WAFS that supports hpc-stack so I can stick with the EMC_post hash I have now and then update to a new EMC_post version when WAFS is ready.

> I saw the version of ufs-weather-model still uses old upp library 10.0.0. If @junwang-noaa provides you an up-to-date version, I might give you a new version of EMC_post for matching.

Okie dokie... @junwang-noaa do you recommend a newer hash of ufs-weather-model? Thanks!

**Comment by GeorgeGayno-NOAA** [comment]
> > @KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?
> 
> Thank you for catching my mistake. That hash should have been 9d9dd23e37a51fcc5ff6b7499b834c85ab32e5f3 (develop branch with beta stack version). I also tested with your feature/stack branch. I'd like to submit this PR in the next week. Thanks!

Would you prefer UFS_UTILS use the 'official' stack? I can accelerate the merge.

**Comment by KateFriedman-NOAA** [comment]
> Would you prefer UFS_UTILS use the 'official' stack? I can accelerate the merge.

Yes please, thanks @GeorgeGayno-NOAA !

File: parm/config/config.fv3 [hunk 1]
```diff
@@ -110,9 +110,9 @@ case $case_in in
         export layout_x_gfs=16
         export layout_y_gfs=16
         export npe_wav=140
-        export npe_wav_gfs=440
+        export npe_wav_gfs=140
         export nth_fv3=4
-        export nth_fv3_gfs=7
+        export nth_fv3_gfs=2
```
Comment (WalterKolczynski-NOAA) [inline parm/config/config.fv3]: These seem like big resource changes.

File: parm/config/config.resources [hunk 2]
```diff
@@ -168,8 +168,10 @@ elif [ $step = "fcst" ]; then
     export npe_fcst=$(echo "$layout_x * $layout_y * 6" | bc)
     export npe_fcst_gfs=$(echo "$layout_x_gfs * $layout_y_gfs * 6" | bc)
     export nth_fcst=${nth_fv3:-2}
+    export nth_fcst_gfs=${nth_fv3_gfs:-2}
     if [[ "$CASE" == "C384" ]]; then nth_fcst=4; fi
     export npe_node_fcst=$(echo "$npe_node_max / $nth_fcst" | bc)
+    export npe_node_fcst_gfs=$(echo "$npe_node_max / $nth_fcst_gfs" | bc)
```
Comment (WalterKolczynski-NOAA) [inline parm/config/config.resources]: Why this second specification for resources?

File: ush/rocoto/workflow_utils.py [hunk 3]
```diff
@@ -295,7 +295,10 @@ def get_resources(machine, cfg, task, reservation, cdump='gdas'):
         ppn = cfg['npe_node_%s' % ltask]
 
     if machine in [ 'WCOSS_DELL_P3', 'HERA', 'ORION']:
-        threads = cfg['nth_%s' % ltask]
+        if cdump in ['gfs'] and 'nth_%s_gfs' % task in cfg.keys():
+            threads = cfg['nth_%s_gfs' % ltask]
+        else:
+            threads = cfg['nth_%s' % ltask]
```
Comment (WalterKolczynski-NOAA) [inline ush/rocoto/workflow_utils.py]: More with the second resource specification.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Why are resource changes even included in a PR migrating to the stack?

File: parm/config/config.fv3 (hunk 1 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.fv3]: reverted

File: parm/config/config.resources (hunk 2 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.resources]: reverted

File: ush/rocoto/workflow_utils.py (hunk 3 above)
Comment (KateFriedman-NOAA) [reply ush/rocoto/workflow_utils.py]: reverted

**Comment by KateFriedman-NOAA** [comment]
@junwang-noaa @yangfanglin @RussTreadon-NOAA @WenMeng-NOAA @YaliMao-NOAA  This PR is in pre-commit testing and is ready to go into global-workflow develop. Please confirm your respective component hash versions mentioned in [first comment of PR](https://github.com/NOAA-EMC/global-workflow/pull/316#issue-621133376) are still good or if there are more recent hashes/tags I should update to and do a final retest with. Thanks!

**Comment by RussTreadon-NOAA** [comment]
NOAA-EMC/GSI at [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6) is several months old (3/9/2021 commit).  The NOAA-EMC/GSI master was updated this morning (6/21/2021) at [a67b816](https://github.com/NOAA-EMC/GSI/commit/a67b816a448d23d2010e9c39741af65a1c56b8f1).   I assume developers will want to use the head of NOAA-EMC/GSI for development.   While this PR could be updated to a67b816, this commit will also become obsolete with time.   NOAA-EMC/GSI development continues moving forward.   Developers can always jump to the head of NOAA-EMC/GSI by editing their local copy of `sorc/checkout.sh`.   

My preference is to work with the most recent versions.   That said, if updating DA in this PR to a67b816 requires rerunning global parallels on multiple machines before this PR may be closed, this may not be a staff and compute cost we want to incur at present.  We may opt to close this PR with DA at 9c1fc15 and open new PRs to bring the workflow closer to the head of DA development.   

These comments may also apply to other workflow components (model, UPP, verification, etc)

**Comment by WenMeng-NOAA** [comment]
@KateFriedman-NOAA I looked at the harsh version of EMC_post and upp lib version at ufs-weather-model (fv3gfs). The offline post and inline post are matched.

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA I looked at the harsh version of EMC_post and upp lib version at ufs-weather-model (fv3gfs). The offline post and inline post are matched.

Great, thanks for checking and confirming @WenMeng-NOAA !

**Comment by KateFriedman-NOAA** [comment]
> NOAA-EMC/GSI at [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6) is several months old (3/9/2021 commit). The NOAA-EMC/GSI master was updated this morning (6/21/2021) at [a67b816](https://github.com/NOAA-EMC/GSI/commit/a67b816a448d23d2010e9c39741af65a1c56b8f1). I assume developers will want to use the head of NOAA-EMC/GSI for development. While this PR could be updated to a67b816, this commit will also become obsolete with time. NOAA-EMC/GSI development continues moving forward. Developers can always jump to the head of NOAA-EMC/GSI by editing their local copy of `sorc/checkout.sh`.
> 
> My preference is to work with the most recent versions. That said, if updating DA in this PR to a67b816 requires rerunning global parallels on multiple machines before this PR may be closed, this may not be a staff and compute cost we want to incur at present. We may opt to close this PR with DA at 9c1fc15 and open new PRs to bring the workflow closer to the head of DA development.

Thanks @RussTreadon-NOAA , I agree with your points and will use the hash slatted for this commit but plan to roll in newer GSI hashes/tags in the near future. Users can update their versions manually for now if they want the latest GSI hashes. Updating single components moving forward will be easier than this initial update to hpc-stack versions.

**Comment by KateFriedman-NOAA** [comment]
@WalterKolczynski-NOAA I believe your review changes have been resolved now. Please rereview and let me know of any other questions or requested changes. Thanks!

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**